### DISCUSSION TITLE
Yarn v3, auth, self hosted for ADO with private repo issues

### DISCUSSION THREAD

**Discussion by namrog84** [renovatebot/renovate#12902] [Request Help] [answered]
I've been setting up renovatebot in a bunch of different areas, npm and yarnv1 without too much issue now for last couple of weeks. 
Got a good handle on many of the things.  However I finally decided to start tackling our Yarn V3 packages.

and running into auth issues.  I am probably overlooking something quite simple.  

```
module.exports = {
  platform: 'azure',
  endpoint: 'https://dev.azure.com/microsoft/',
  token: process.env.TOKEN,
  allowPostUpgradeCommandTemplating: true,
  allowScripts: true,
  trustlevel: 'high',
  hostRules: [
    {
      hostType: 'npm',
      matchHost: 'https://microsoft.pkgs.visualstudio.com/_packaging/gaming/npm/registry/',
      username:"microsoft",
      password: process.env.TOKEN,
    },
  ],
  repositories: [process.env.RENOVATE_TARGET_REPO],
};
```

We have a variety of packages setup inside
/src/web/packages/*    (sits about 3 package.json in each one)
then a root one src/web/package.json as we are using workspaces, so there is only the single src/web/yarn.lock file

Inside our /src/web/.yarn.yml
```
nodeLinker: node-modules
npmAlwaysAuth: true
npmRegistryServer: "https://microsoft.pkgs.visualstudio.com/_packaging/gaming/npm/registry/"
plugins:
  // misc internal plugins (I can share more if relevant)
yarnPath: .yarn/releases/yarn-3.1.0.cjs
```

It successfully creates a branch, and updates package.json file, but it fails to update the yarn.lock file

e.g.
```
DEBUG: rawExec err (repository=Xbox/GS, branch=renovate/babylonjs-core-5.x)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "yarn install --mode=update-lockfile",
         "stdout": "➤ YN0000: ┌ Resolution step\n➤ YN0041: │ tslib@npm:^2.3.1: Invalid authentication (as an anonymous user)\n➤ YN0000: └ Completed in 1s 102ms\n➤ YN0000: Failed with errors in 1s 105ms\n",
         "stderr": "",
         "message": "Command failed: yarn install --mode=update-lockfile\n",
         "stack": "Error: Command failed: yarn install --mode=update-lockfile\n\n    at ChildProcess.exithandler (node:child_process:397:12)\n    at ChildProcess.emit (node:events:390:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1064:16)\n    at Socket.<anonymous> (node:internal/child_process:450:11)\n    at Socket.emit (node:events:390:28)\n    at Socket.emit (node:domain:475:12)\n    at Pipe.<anonymous> (node:net:687:12)"
       }
DEBUG: lock file error (repository=Xbox/GS, branch=renovate/babylonjs-core-5.x)
       "err": {
         ... same as above
       },
       "type": "yarn"
DEBUG: No updated lock files in branch (repository=Xbox/GS, branch=renovate/babylonjs-core-5.x)
```

I've been trying a bunch of different things.  While the original above works fine for npm and yarnv1 packages,  yarnv3 is unhappy

I did see someone mentioned adding encrypted to the password
`encrypted: { password: process.env.TOKEN},` 

Which causes a bunch of 401s to start happening
e.g. ` "message": "Response code 401 (Unauthorized)",   "stack": "HTTPError: Response code 401 (Unauthorized)\`
The original did not ever say '401' 

Not sure if 401 unauthorized is closer or further from solution than Invalid authentication?

I've also tried using token:... and encrypted: {token:...   regular token also gives me the standard anonymous user originally above, and encrypted one gives the same 401 issue.

**Comment by rarkins** [comment, marked as answer]
What do you need src/web/.yarn.yml to look like once the username/password is added into it by Renovate before calling Yarn? Alternatively how do you run it locally? E.g. a .yarn.yml in your home directory?

**Reply by namrog84** [reply]
I think/thought yarn v2/v3 uses a user level auth now instead of folder specific one like npm/yarnv1 did

Locally my c:/users/namrog/.yarnrc.yml looks like this in a successful auth scenario for local dev. 

```
npmRegistries:
  "https://microsoft.pkgs.visualstudio.com/_packaging/gaming/npm/":
    npmAlwaysAuth: true
    npmAuthIdent: hashsecretremoved=
  "https://microsoft.pkgs.visualstudio.com/_packaging/gaming/npm/registry/":
    npmAlwaysAuth: true
    npmAuthIdent: hashsecretremoved=
```

Which I guess means I'll try adding a secondary matchHost for this other path without the /registry/ and see if it works. 

Maybe it needs both?

**Reply by namrog84** [reply]
Adding both matchHosts appears to be working for in my testing environment. I'll try integrating it tomorrow in the main repo which combines all the things together.


```
    {
      hostType: 'npm',
      matchHost: 'https://microsoft.pkgs.visualstudio.com/_packaging/gaming/npm/registry/',
      username:"microsoft",
      password: process.env.TOKEN
    },
    {
      hostType: 'npm',
      matchHost: 'https://microsoft.pkgs.visualstudio.com/_packaging/gaming/npm/',
      username:"microsoft",
      password: process.env.TOKEN
    },
```

**Reply by namrog84** [reply]
So it sorta works for some of them but not all of them. I am not entirely sure why.
There is still 1 case that I am seeing the anonymous user issue. And it's still part of the same workspace.  so its really odd that it works for some but not all.


I know we typically handle auth by a custom 'yarn auth' plugin, we run before yarn install, and for our own CI machines we often do a 
```
export AUTH=$(system.accessToken)
yarn auth
yarn install
```

And in renovate, while I can get the env variable in the system beforehand, there is no 'preinstall' script as far as I can tell? 


1. Does it matter we have a monorepo with both yarn v1 and yarn v3 usage in same repo?  Like that's not gonna prevent it from behaving correctly is it?

**Reply by namrog84** [reply]
I managed to run our personal yarn auth before we run npx renovate and because yarn v3 is user wide auth and not repo/folder specific, it was good enough to unblock us.

Case closed for now!  Thanks!
